Added initial code for Auth.UI.WinUI3 and WinUI3 sample

anovik opened the pull request.

anovik committed:
Initial code for WinUI3 library and sample.

bezysoftware commented on src/Auth.UI.WinUI3/tools/VisualStudioToolsManifest.xml:
```diff
@@ -0,0 +1,7 @@
+﻿<FileList>
+  <File Reference="Firebase.Auth.UI.UWP.dll">
```
UWP should be WinUI3?

anovik replied:
Fixed

bezysoftware commented on src/Auth.UI.WinUI3/Themes/Generic.xaml:
```diff
@@ -0,0 +1,35 @@
+    </Style>
+
+    <ResourceDictionary.MergedDictionaries>
+        <ResourceDictionary Source="ms-appx:///Firebase.Auth.UI.UWP/Themes/FirebaseUIControl.xaml" />
```
UWP should be WinUI3?

anovik replied:
Fixed

bezysoftware commented on src/Auth.UI.WinUI3/Auth.UI.WinUI3.csproj:
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net6.0-windows10.0.19041.0</TargetFramework>
+    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
```
How did you choose these versions, are they reasonable? (I'm not so familiar with Win10 version, forgive my dumb question :) )

anovik replied:
They are the default values for WinUI3 now. 
It means that it will work on all versions of Windows 11 and Windows 10 starting from  Windows 10, version 1809 - that includes most of Windows 10 systems.

bezysoftware reviewed (changes requested):
Can you also please quickly summarize the difference between UWP (is that no longer supported by MS?) and WinUI3? Can WinUI3 be used in UWP / WPF / Forms apps, or only somewhere. Thanks

anovik commented:
UWP is officially still supported by Microsoft but they don't make any new changes since 2016. So it is quite outdated. 

WinUI3 is a new native Windows user interface platform, that goes with Windows App SDK. So it is a new alternative to WPF, I would say.

anovik committed:
Fixed some namespaces and resource names.

bezysoftware commented:
I am not able to run the sample on VS 2022. There are several issues:
- The calls to UI Dispatcher do not work - it seems you should be using DispatcherQueue to run stuff on UI thread?
- When user is anonymous, you should still see the login page (line 64 in app.xaml.cs)
- When login page should be displayed it throws an error saying `XAML parsing failed.`
Can you take a look?

bezysoftware commented:
Also we should probably note somewhere that 'Single-project MSIX Packaging Tools for VS 2022' needs to be installed to run the sample, I'll add it to readme once this is merged

anovik committed:
Fixed Dispatcher and resource dictionary issues.

anovik committed:
Updated message boxes and dispatcher queue to WinUI3.
Temporarily commented ProgressBar.

anovik commented:
@bezysoftware I fixed the issues with resources and the dispatcher queue. Also updated message boxes. 

> Also we should probably note somewhere that 'Single-project MSIX Packaging Tools for VS 2022' needs to be installed to run the sample, I'll add it to readme once this is merged

That's correct.

anovik commented:
The two issues which are still here for WinUI3:

1. I had to comment ProgressBar, there is some problem with it not in the code but in WinUI3. Adding it to XAML leads to an exception.
2. For Google WebAuthenticationBroker.AuthenticateAsync doesn't display any window as it should.

bezysoftware commented:
It seem neither of the auth providers which use WebAuthenticationBroker work - no window is displayed. 
Seems WAB is only supported on UWP, nowhere else: https://github.com/microsoft/WindowsAppSDK/issues/441 so a different solution needs to be used - see a "fake" WAB in the WPF implementation - it uses a custom Window with WebView2

anovik committed:
WebAuthenticationBroker implementation for WinUi3.

anovik commented:
@bezysoftware I've added the initial implementation of custom WebAuthenticationBroker for WinUI3. For now, I have some initialization problems with `WebView.EnsureCoreWebView2Async()` on my machine, it throws a catastrophic failure COM exception. If you have time, could you please check it also? Will continue investigating this also.

bezysoftware commented:
Can you try running this gallery sample? https://github.com/microsoft/WinUI-Gallery It's also published in Microsoft Store to simply install. It contains WebView2 sample, for me it works (page loads), but in Firebase project it still doesn't load

anovik committed:
Added full-trust permission to WinUI3 sample.

anovik commented:
@bezysoftware Finally, I solved the problem with the stupid WebView2 (I hate it!) Added full-trust permission to winui3 sample and it works. 
Could you please check it from your side? Everything should be now functional.

bezysoftware commented on samples/WinUI3/App.xaml.cs:
```diff
@@ -0,0 +1,105 @@
+                    }
+                    else if (e.User.IsAnonymous)
+                    {
+                        (Window.Content as Frame).Navigate(typeof(MainPage));
```
Anonymous user should still be taken to LoginPage

anovik replied:
Fixed

bezysoftware commented on samples/WinUI3/App.xaml.cs:
```diff
@@ -0,0 +1,105 @@
+
+        private void AuthStateChanged(object sender, UserEventArgs e)
+        {
+            var dispatcherQueue = DispatcherQueue.GetForCurrentThread();
```
This returns null for me sometimes, would it be better to use the `Window?.DispatcherQueue` as we need to run this code on the main thread?

anovik replied:
I don't have this problem but still I saved dispatcher queue from OnLaunched where it should never be null and in AuthStateChanged I use this variable. Could you please check the problem is solved?

bezysoftware commented on samples/WinUI3/Properties/launchSettings.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "profiles": {
+    "Auth.UWP.Sample (Package)": {
```
Can you rename these profiles to `WinUI3` ?

anovik replied:
Fixed

anovik committed:
Fixed launchSettings.json

anovik committed:
Added nuget information for Auth.UI.WinUI3.

anovik commented:
>there is also the stuff for nuget package which is missing in `csproj`, you can either add it yourself (see the WPF csproj for example) or I'll add it after merge.

I've added nuget information for winui3.

anovik committed:
dispatcherQueue and Anonymous user changes.

anovik commented:
@bezysoftware It seems the failure doesn't refer to the WinUI3 part, just tests require NET Core 3.1 and it is absent in the environment build. It is probably worth upgrading tests to NET 6.

anovik commented:
@bezysoftware Any update regarding this PR?

bezysoftware commented:
I've been busy with other stuff, I should hopefully have some time next week

bezysoftware commented:
I upgraded the tests, merged your PR and made a few finishing touches, looks like the solution is building now. I'll test a bit tomorrow and deploy to nuget after that, will let you know here.
Thanks for the PR!

bezysoftware commented:
Nuget is published: https://www.nuget.org/packages/FirebaseAuthentication.WinUI3 🎉 
Thanks again! @anovik
